pkg/controller/instanceset/instance_util.go

```diff
@@ -263,7 +289,71 @@ func GenerateInstanceNames(parentName, templateName string,
 			}
 		}
 	}
-	return instanceNameList, ordinal
+	return instanceNameList, nil
+}
+
+// GenerateInstanceNamesWithOrdinalList generates instance names based on ordinalList and offlineInstances.
+func GenerateInstanceNamesWithOrdinalList(parentName, templateName string,
+	replicas int32, offlineInstances []string, ordinalList []int32) ([]string, error) {
+	var instanceNameList []string
+	usedNames := sets.New(offlineInstances...)
+	for _, ordinal := range ordinalList {
+		var name string
+		if len(templateName) == 0 {
+			name = fmt.Sprintf("%s-%d", parentName, ordinal)
+		} else {
+			name = fmt.Sprintf("%s-%s-%d", parentName, templateName, ordinal)
+		}
+		if usedNames.Has(name) {
+			continue
+		}
+		instanceNameList = append(instanceNameList, name)
+	}
+	if int32(len(instanceNameList)) != replicas {
+		errorMessage := fmt.Sprintf("for template '%s', expected %d instance names but generated %d: [%s]",
+			templateName, replicas, len(instanceNameList), strings.Join(instanceNameList, ", "))
+		return instanceNameList, fmt.Errorf(errorMessage)
+	}
+	return instanceNameList, nil
+}
+
+func GetOrdinalListByTemplateName(its *workloads.InstanceSet, templateName string) ([]int32, error) {
+	ordinals, err := GetOrdinalsByTemplateName(its, templateName)
+	if err != nil {
+		return nil, err
+	}
+	return ConvertOrdinalsToOrdinalList(ordinals)
+}
+
+func GetOrdinalsByTemplateName(its *workloads.InstanceSet, templateName string) (workloads.Ordinals, error) {
+	if templateName == "" {
+		return its.Spec.DefaultTemplateOrdinals, nil
+	}
+	for _, template := range its.Spec.Instances {
+		if template.Name == templateName {
+			return template.Ordinals, nil
+		}
+	}
+	return workloads.Ordinals{}, fmt.Errorf("template %s not found", templateName)
+}
+
+func ConvertOrdinalsToOrdinalList(ordinals workloads.Ordinals) ([]int32, error) {
+	ordinalList := sets.New(ordinals.Discrete...)
+	for _, item := range ordinals.Ranges {
+		start := item.Start
+		end := item.End
+
+		if start > end {
+			return nil, fmt.Errorf("range's end(%v) must >= start(%v)", end, start)
+		}
+
+		for ordinal := start; ordinal <= end; ordinal++ {
```

Comment:
First look up the ordinal in set to judge if there are overlapping regions in ranges and diacretes，if existed，log warning messages for further debugging